src/vectors.S

```diff
@@ -0,0 +1,92 @@
+.section .text.init
+.global _run_guest_vector,_guest_exit_vector
+
+// t6 -> cached returned value
+// t5 -> cached sstatus
+// restore vectors for guest
+// i.e. memory to main state
+.align 2
+_run_guest_vector:
+    csrr t5, sstatus
+
+    // if vs != 0, restore
+    // enable vectors
+
+    ld t1, ({guest_vl})(a0)
+    ld t2, ({guest_vtype})(a0)
+    vsetvl t0,t1,t2
+
+    // check that vtype is ok?
+
+    ld t0, ({guest_vstart})(a0)
+    csrw vstart, t0
+
+    ld t0, ({guest_vcsr})(a0)
+    csrw vcsr, t0
+    
+
+
+    addi t3, a0, {guest_v0}
+    vl8r.v  v0, (t3)
+    addi t3, a0, {guest_v8}
+    vl8r.v  v8, (t3)
+    addi t3, a0, {guest_v16}
+    vl8r.v  v16, (t3)
+    addi t3, a0, {guest_v24}
+    vl8r.v  v24, (t3)
+
+    csrw sstatus, t5
+    ret
+
+// t6 -> cached returned value
+// t1 -> cached sstatus
+// save vectors for guest
+// i.e. from guest to mem
+.align 2
+_guest_exit_vector:
+    csrr  t1, sstatus
+    li    t0, {sstatus_vs_dirty}
+    and   t2, t1, t0
+    bne   t2, t0, _guest_exit_vector_return
+    not   t0, t0
+    and   t1, t1, t0
+    li    t0, {sstatus_vs_clean}
+    or    t1, t1, t0
+
+    csrrs  t4, vcsr, zero
+    sd t4, {guest_vcsr}(a0)
+
+    csrrs t4, vstart, zero
+    sd t4, {guest_vstart}(a0)
+
+    csrrs t4, vtype, zero
+    sd t4, {guest_vtype}(a0)
+
+    csrrs t4, vl, zero
+    sd t4, {guest_vl}(a0)
+
+    addi t3, a0, {guest_v0}
+    vs8r.v  v0, (t3)
+    addi t3, a0, {guest_v8}
+    vs8r.v  v8, (t3)
+    addi t3, a0, {guest_v16}
+    vs8r.v  v16, (t3)
+    addi t3, a0, {guest_v24}
+    vs8r.v  v24, (t3)
+
+    // for debugging only, clear out vector registers 
+    // to be sure they are restored correctly
+    // TODO: remove this before going into production
+    vmv.v.i v0, 0
+    vmv.v.i v4, 0
+    vmv.v.i v8, 0
+    vmv.v.i v12, 0
+    vmv.v.i v16, 0
+    vmv.v.i v20, 0
+    vmv.v.i v24, 0
+    vmv.v.i v28, 0
```

Comment:
Remove?

Reply:
Not yet. I think it's best to wait a while until we do. It makes it very very clear if there is a problem. Without doing it, and problem with the vector save/restore might not show up. I would argue that we should keep it around for a little while.

Reply:
Can the clearing here be replaced by a test with the following sequence?

Tellus sets vector reg
run guest vm
guest vm set vector regs to different values
guest vm exits
tellus resumes and checks that its register values have been restored.

Reply:
Tellus is pretty much vector agnostic. The issue is that we only have one guest vm running. Once we start doing multiple guest vm's, this can go away.

What happens is that the vector registers get set, and no other "process" uses them, so the never get cleared and we can't check if they were restored correctly.  I think we can just leave this in until we are running multiple guest vm's.

Reply:
> The issue is that we only have one guest vm running

Note that the "host" (Tellus) is itself a VM so the procedure Dylan described would exercise the context switch since Salus must save/restore vector state when switching between Tellus and guestvm.